PropertyExpression.java

```diff
@@ -0,0 +1,99 @@
+        this.owner = owner;
+    }
+    
+    /**
+     * If you're property only relies on one simple method, you can define that method here
+     * using a {@link Function}. This function will be applied at the {@link #getValues(TriggerContext)} method
+     * of this class.
+     * @return the function that needs to be applied in order to get the correct values.
+     */
+    public Function<O[], T[]> getPropertyFunction() {
+        return null;
+    }
+
+    /**
+     * There are 2 kinds of possession:
+     * <ul>
+     *     <li><b>Genitive:</b>Mwexim's book</li>
+     *     <li><b>Regular:</b>book of Mwexim</li>
+     * </ul>
+     * Because both have different patterns and the order of the expressions can be important,
+     * you can use this method to check if the pattern is the Genitive form. If it's not, it's the regular form.
+     * @param matchedPattern the matched pattern of a property
+     * @return whether this pattern was the genitive form (true) or the regular form (false)
+     */
+    public static boolean isGenitive(int matchedPattern) {
+        return matchedPattern == 0;
+    }
+
+    /**
+     * This {@code init()} method overrides the default method and does the initialization by itself.
+     * Most of the time, you'll still want to override this, because the only thing it does is
+     * getting the first expression (because there is only one) and changing the {@link #owner} field
+     * accordingly.
+     * @param expressions an array of expressions representing all the expressions that are being passed
+     *                    to this syntax element. As opposed to Buzzle, elements of this array can't be {@code null}.
+     * @param matchedPattern the index of the pattern that was successfully matched. It corresponds to the order of
+     *                       the syntaxes in registration
+     * @param parseContext an object containing additional information about the parsing of this syntax element, like
+     *                    regex matches and parse marks
+     * @return whether the initialization was successful or not.
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public boolean init(Expression<?>[] expressions, int matchedPattern, ParseContext parseContext) {
+        setOwner((Expression<O>) expressions[0]);
+        return getOwner() != null;
+    }
+
+    /**
+     * A simple default method that will apply your basic function on the {@link #owner} of this property.
+     * It checks for nullity and returns the appropriate values.
+     *
+     * @param ctx the event
+     * @return the values of this property after applying the {@link #getPropertyFunction()} function on the owner.
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public T[] getValues(TriggerContext ctx) {
+        O[] objs = getOwner().getValues(ctx);
+        if (objs == null) return (T[]) new Object[0];
+        if (getPropertyFunction() == null) return (T[]) new Object[0];

```

Comment:
It doesn't really make sense for this check to just return null (well, the equivalent thereof). If someone overrides getPropertyFunction() in any reasonable fashion, the result would be non-null. And if they don't, it would be nonsensical not to override getValues() instead.

So I think this should instead throw an UnsupportedOperationException, and then it should be mentioned in the specification that if getPropertyFunction() isn't overridden, then getValues() must be.